lattice-controller/src/client_cache.rs

```diff
@@ -0,0 +1,248 @@
+                )))
+            }
+        }
+    }
+
+    async fn store_client(&self, lattice_id: &str, client: Client) {
+        let mut conns = self.clients.write().await;
+        conns.insert(lattice_id.to_string(), client);
+    }
+
+    async fn record_access(&self, lattice_id: &str) {
+        let mut meta = self.meta.write().await;
+        meta.entry(lattice_id.to_string()).and_modify(|e| e.touch());
+    }
+}
+
+async fn create_client(
+    config: &ConnectionConfig,
+) -> RpcResult<wasmcloud_control_interface::Client> {
+    let timeout = Duration::from_millis(config.timeout_ms);
+    let auction_timeout = Duration::from_millis(config.auction_timeout_ms);
+    let lattice_prefix = config.lattice_prefix.clone();
+    let conn = connect(config).await?;
+    let client = wasmcloud_control_interface::Client::new(
+        conn,
+        Some(lattice_prefix),
+        timeout,
+        auction_timeout,
+    );
+    Ok(client)
+}
+
+/// Create a new nats connection
+async fn connect(cfg: &ConnectionConfig) -> RpcResult<async_nats::Client> {
+    let cfg = cfg.clone();
+    let opts = match (cfg.auth_jwt, cfg.auth_seed) {
+        (Some(jwt), Some(seed)) => {
+            let key_pair = std::sync::Arc::new(
+                KeyPair::from_seed(&seed)
+                    .map_err(|e| RpcError::ProviderInit(format!("key init: {}", e)))?,
+            );
+            async_nats::ConnectOptions::with_jwt(jwt, move |nonce| {
+                let key_pair = key_pair.clone();
+                async move { key_pair.sign(&nonce).map_err(async_nats::AuthError::new) }
+            })
+        }
+        (None, None) => async_nats::ConnectOptions::default(),
+        _ => {
+            return Err(RpcError::InvalidParameter(
+                "must provide both jwt and seed for jwt authentication".into(),
+            ));
+        }
+    };
+    if cfg.cluster_uris.is_empty() {
+        return Err(RpcError::NotInitialized(
+            "No NATS URIs supplied".to_string(),
+        ));
+    }
+
+    let url = cfg.cluster_uris.get(0).unwrap();
+
+    let conn = opts
+        .event_callback(|event| async move {
+            // lattice prefix/ID will already be on the span from earlier calls
+            match event {
+                async_nats::Event::Disconnect => debug!("NATS client disconnected"),
+                async_nats::Event::Reconnect => debug!("NATS client reconnected"),
+                async_nats::Event::ClientError(err) => debug!("NATS client error occured: {}", err),
+                other => debug!("NATS client other event occurred: {}", other),
+            }
+        })
+        .connect(url)
+        .await
+        .map_err(|e| RpcError::ProviderInit(format!("Nats connection to {}: {}", url, e)))?;
+
+    Ok(conn)
+}
+
+/// Discovers a list of expired (access time within grace period) connections
+/// and then removes them from the cache.
+async fn evacuate_cache(
+    m: Arc<RwLock<HashMap<String, ClientMetadata>>>,
```

Comment:
I would have preferred dashmap over Hashmap to reduce lock contention, but I see you left a todo to evaluate. 

Please add two more TODOs:
-  evaluate cache crate like https://crates.io/crates/cached which I think would reduce and simplify the code in this crate. 
- increase number of unit tests (not necessary if we switch to `cached` or similar crate if it already has a reasonable test suite).

Reply:
I tried dashmap and couldn't get it to work after 3 hours. this is why there's a TODO at the top of this file to explore dashmap further.